compiler/rustc_mir_build/src/thir/pattern/_match.rs

```diff
@@ -522,18 +536,65 @@ impl<'p, 'tcx> Matrix<'p, 'tcx> {
                 self.push(row)
             }
         } else {
-            self.0.push(row);
+            if self.patterns.is_empty() {
```

Comment:
Please put this code into a separate function and document it in much more detail. This module is already super complex and adding a fast path to it that does not change behaviour does not help with that complexity.